doc/sonic-sub-port-intf-hld.md

```diff
@@ -0,0 +1,493 @@
+
+We use iproute2 package to manage host sub port interfaces.
+Specifically, we use `ip link add link <parent_port_name> name <subif_name> type vlan id <vlan_id>` to create a host sub port interface.
+This command implies the dependancy that a parent host interface must be created before the creation of a host sub port interface.
+At creation, a host sub port interface is always set to admin status up.
+
+Example:
+```
+ip link add link Ethernet64 name Ethernet64.10 type vlan id 10
+ip link set Ethernet64.10 mtu 9100
+ip link set Ethernet64.10 up
+```
+```
+ip link del Ethernet64.10
+```
+
+We use `ip address` and `ip -6 address` to add and remove ip adresses on a host sub port interface.
+
+Example:
+```
+ip address add 192.168.0.1/24 dev Ethernet64.10
+```
+
+### 2.5.2 Route, neighbor subsystems
+
+Once the host sub port interfaces are properly set, route and neighbor subsystems should function properly on sub port interfaces.
+fpmsyncd should receive route add/del updates on sub port interfaces from zebra over TCP socket port 2620. These updates are received in the format of netlink messages.
+neighsyncd should receive neigh add/del netlink messages on sub port interfaces from its subscription to the neighbor event notification group RTNLGRP_NEIGH.
+
+Internally, a sub port interface is represented as a Port object to be perceived seamlessly by NeighOrch and RouteOrch to create neighor and route entries, respectively, on it.
+
+
+# 3 Event flow diagrams
+## 3.1 Sub port interface creation
+![](https://github.com/wendani/SONiC/blob/sub_port_master/images/sub_interface_hld/sub_intf_creation_flow.png)
+
+## 3.2 Sub port interface runtime admin status change
+![](https://github.com/wendani/SONiC/blob/sub_port_master/images/sub_interface_hld/sub_intf_set_mtu_flow.png)
+
+## 3.3 Sub port interface removal
+![](https://github.com/wendani/SONiC/blob/sub_port_master/images/sub_interface_hld/sub_intf_removal_flow.png)
+
+# 4 CLIs
+## 4.1 Config commands
+`subinterface` command category is introduced to the `config` command.
+
+```
+Usage: config [OPTIONS] COMMAND [ARGS]...
+
+  SONiC command line - 'config' command
+
+Options:
+  --help  Show this message and exit.
+
+Commands:
+  ...
+  subinterface           Sub-port-interface-related configuration tasks
+```
+
+`add` and `del` commands are supported on a sub port interface.
+```
+Usage: config subinterface [OPTIONS] COMMAND [ARGS]...
+
+  Sub-port-interface-related configuration tasks
+
+Options:
+  --help    Show this message and exit.
+
+Commands:
+  add       Add a sub port interface
+  del       Remove a sub port interface
+  ip        Add or remove an IP address
+```
+```
+Usage: config subinterface add [OPTIONS] <sub_port_interface_name>
+```
+```
+Usage: config subinterface del [OPTIONS] <sub_port_interface_name>
+```
+
+Once a sub port interface is added, `add` and `del` ip address commands are supported on it.
+```
+Usage: config subinterface ip [OPTIONS] COMMAND [ARGS]...
```

Comment:
Please use existing CLI to configure IP address for an interface. It also takes care of any default-VRF handling. Suggest to have one CLI to configure ip address for an interface.